File: dsa/examples/use-cases/firmware-signing/verify_firmware.py
#!/usr/bin/env python3
"""
Post-Quantum Firmware Signature Verification Tool

Verify firmware signatures and check rollback protection.

Features:
- Cryptographic signature verification
- Firmware hash integrity check
- Rollback protection (version checking)
- Device compatibility verification

Usage:
    python verify_firmware.py --key public.key --firmware firmware.bin [options]

Exit codes:
    0 - Signature valid, firmware safe to install
    1 - Verification failed (signature invalid, hash mismatch, etc.)
    2 - Input error (file not found, etc.)
    3 - Rollback protection triggered (firmware version too old)
    4 - Device compatibility error
"""

import argparse
import hashlib
import json
import sys
from datetime import datetime
from pathlib import Path
from typing import Optional

# Try to import from installed package first
try:
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )
except ImportError:
    sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "src" / "py"))
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )


# Supported algorithms
ALGORITHMS = {
    "mldsa44": {"class": MLDSA44, "name": "ML-DSA-44"},
    "mldsa65": {"class": MLDSA65, "name": "ML-DSA-65"},
    "mldsa87": {"class": MLDSA87, "name": "ML-DSA-87"},
    "slh-shake-128f": {"class": SLHDSA_SHAKE_128f, "name": "SLH-DSA-SHAKE-128f"},
    "slh-shake-128s": {"class": SLHDSA_SHAKE_128s, "name": "SLH-DSA-SHAKE-128s"},
    "slh-shake-192f": {"class": SLHDSA_SHAKE_192f, "name": "SLH-DSA-SHAKE-192f"},
    "slh-shake-256f": {"class": SLHDSA_SHAKE_256f, "name": "SLH-DSA-SHAKE-256f"},
}


class VerificationError(Exception):
    """Raised when signature verification fails."""
    pass


class IntegrityError(Exception):
    """Raised when firmware integrity check fails."""
    pass


class RollbackError(Exception):
    """Raised when rollback protection is triggered."""
    pass


class CompatibilityError(Exception):
    """Raised when device compatibility check fails."""
    pass


def compute_firmware_hash(filepath: Path) -> dict:
    """Compute hashes of firmware image."""
    sha256 = hashlib.sha256()
    sha512 = hashlib.sha512()

    with open(filepath, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            sha256.update(chunk)
            sha512.update(chunk)

    return {
        "sha256": sha256.hexdigest(),
        "sha512": sha512.hexdigest(),
    }


def load_public_key(key_path: Path) -> bytes:
    """Load public key from file."""
    key_data = key_path.read_bytes()

    if key_data.startswith(b'{'):
        key_json = json.loads(key_data)
        key_hex = key_json.get("key", key_json.get("public_key", ""))
        return bytes.fromhex(key_hex)

    return key_data


def load_manifest(manifest_path: Path) -> dict:
    """Load and parse firmware manifest."""
    try:
        with open(manifest_path, "r") as f:
            return json.load(f)
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid manifest format: {e}")


def verify_firmware_integrity(filepath: Path, manifest: dict) -> None:
    """
    Verify firmware file integrity against manifest.

    Raises:
        IntegrityError: If firmware doesn't match manifest
    """
    if not filepath.exists():
        raise IntegrityError(f"Firmware file not found: {filepath}")

    firmware_info = manifest.get("firmware", {})

    # Check file size
    expected_size = firmware_info.get("size")
    actual_size = filepath.stat().st_size
    if expected_size is not None and actual_size != expected_size:
        raise IntegrityError(
            f"Firmware size mismatch: expected {expected_size}, got {actual_size}"
        )

    # Check hashes
    expected_hashes = firmware_info.get("hashes", {})
    actual_hashes = compute_firmware_hash(filepath)

    # Check SHA-256
    if expected_hashes.get("sha256"):
        if actual_hashes["sha256"] != expected_hashes["sha256"]:
            raise IntegrityError(
                f"SHA-256 hash mismatch:\n"
                f"  Expected: {expected_hashes['sha256']}\n"
                f"  Got:      {actual_hashes['sha256']}"
            )

    # Check SHA-512 if present
    if expected_hashes.get("sha512"):
        if actual_hashes["sha512"] != expected_hashes["sha512"]:
            raise IntegrityError(
                f"SHA-512 hash mismatch:\n"
                f"  Expected: {expected_hashes['sha512']}\n"
                f"  Got:      {actual_hashes['sha512']}"
            )


def verify_signature(
    manifest: dict,
    public_key: bytes,
) -> dict:
    """
    Verify cryptographic signature.

    Returns:
        Verification result details

    Raises:
        VerificationError: If signature is invalid
    """
    algo_info = manifest.get("algorithm", {})
    algorithm = algo_info.get("id")

    if algorithm not in ALGORITHMS:
        raise VerificationError(f"Unsupported algorithm: {algorithm}")

    # Get signature
    sig_info = manifest.get("signature", {})
    sig_hex = sig_info.get("value", "")
    signature = bytes.fromhex(sig_hex)

    context_hex = sig_info.get("context", "")
    context = bytes.fromhex(context_hex) if context_hex else b"firmware"

    # Reconstruct signed data
    sign_data = {
        "firmware": manifest.get("firmware"),
        "metadata": manifest.get("metadata"),
        "timestamp": manifest.get("timestamp"),
    }
    sign_message = json.dumps(sign_data, sort_keys=True).encode("utf-8")

    # Verify
    verifier = ALGORITHMS[algorithm]["class"]()

    try:
        is_valid = verifier.verify(public_key, sign_message, signature, ctx=context)
    except Exception as e:
        raise VerificationError(f"Signature verification error: {e}")

    if not is_valid:
        raise VerificationError("Signature verification failed: invalid signature")

    return {
        "algorithm": algo_info.get("name"),
        "security_level": algo_info.get("security_level"),
        "timestamp": manifest.get("timestamp"),
    }


def check_rollback_protection(
    manifest: dict,
    current_version_code: Optional[int] = None,
) -> None:
    """
    Check rollback protection.

    Args:
        manifest: Firmware manifest
        current_version_code: Currently installed firmware version code

    Raises:
        RollbackError: If firmware is older than current version
    """
    if current_version_code is None:
        return  # Skip check if current version not provided

    metadata = manifest.get("metadata", {})
    new_version_code = metadata.get("version_code", 0)

    if new_version_code < current_version_code:
        raise RollbackError(
            f"Rollback protection: new firmware version code ({new_version_code}) "
            f"is older than current ({current_version_code})"
        )


def check_device_compatibility(
    manifest: dict,
    device_type: Optional[str] = None,
    device_model: Optional[str] = None,
) -> None:
    """
    Check device compatibility.

    Raises:
        CompatibilityError: If firmware is not compatible with device
    """
    metadata = manifest.get("metadata", {})

    # Check device type
    if device_type:
        manifest_device_type = metadata.get("device_type", "")
        if manifest_device_type and manifest_device_type != device_type:
            raise CompatibilityError(
                f"Device type mismatch: firmware is for '{manifest_device_type}', "
                f"but device is '{device_type}'"
            )

    # Check compatibility list
    if device_model:
        compatibility = metadata.get("compatibility", [])
        if compatibility and device_model not in compatibility:
            raise CompatibilityError(
                f"Device model '{device_model}' not in compatibility list: {compatibility}"
            )


def format_timestamp(iso_timestamp: str) -> str:
    """Format ISO timestamp for display."""
    try:
        dt = datetime.fromisoformat(iso_timestamp.replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d %H:%M:%S UTC")
    except (ValueError, AttributeError):
        return iso_timestamp or "Unknown"


def main():
    parser = argparse.ArgumentParser(
        description="Verify post-quantum firmware signatures",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Basic verification
  python verify_firmware.py --key public.key --firmware firmware.bin

  # Verify with rollback protection
  python verify_firmware.py --key public.key --firmware firmware.bin \\
      --current-version 1005000

  # Verify device compatibility
  python verify_firmware.py --key public.key --firmware firmware.bin \\
      --device-type "IoT-Sensor-v1" --device-model "Model-A"

Exit codes:
  0 - Verification successful
  1 - Signature/integrity verification failed
  2 - Input error (file not found)
  3 - Rollback protection triggered
  4 - Device compatibility error
        """,
    )

    parser.add_argument(
        "--key", "-k",
        required=True,
        type=Path,
        help="Path to public key file",
    )
    parser.add_argument(
        "--firmware", "-f",
        required=True,
        type=Path,
        help="Path to firmware binary",
    )
    parser.add_argument(
        "--manifest", "-m",
        type=Path,
        help="Path to manifest file (default: <firmware>.fwsig)",
    )

    # Rollback protection
    parser.add_argument(
        "--current-version",
        type=int,
        help="Current firmware version code for rollback protection",
    )

    # Device compatibility
    parser.add_argument(
        "--device-type",
        help="Device type to check compatibility",
    )
    parser.add_argument(
        "--device-model",
        help="Device model to check compatibility",
    )

    parser.add_argument(
        "--quiet", "-q",
        action="store_true",
        help="Suppress output, use exit code only",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Output result as JSON",
    )

    args = parser.parse_args()

    # Validate inputs
    if not args.key.exists():
        if not args.quiet:
            print(f"Error: Public key not found: {args.key}", file=sys.stderr)
        sys.exit(2)

    if not args.firmware.exists():
        if not args.quiet:
            print(f"Error: Firmware not found: {args.firmware}", file=sys.stderr)
        sys.exit(2)

    manifest_path = args.manifest or args.firmware.with_suffix(args.firmware.suffix + ".fwsig")
    if not manifest_path.exists():
        if not args.quiet:
            print(f"Error: Manifest not found: {manifest_path}", file=sys.stderr)
        sys.exit(2)

    # Load inputs
    try:
        public_key = load_public_key(args.key)
        manifest = load_manifest(manifest_path)
    except Exception as e:
        if not args.quiet:
            print(f"Error loading inputs: {e}", file=sys.stderr)
        sys.exit(2)

    result = {"valid": False}

    # Step 1: Verify firmware integrity
    try:
        verify_firmware_integrity(args.firmware, manifest)
    except IntegrityError as e:
        if args.json:
            print(json.dumps({"valid": False, "error": "integrity", "message": str(e)}, indent=2))
        elif not args.quiet:
            print(f"\n[FAILED] Firmware integrity check failed", file=sys.stderr)
            print(f"  {e}", file=sys.stderr)
        sys.exit(1)

    # Step 2: Verify cryptographic signature
    try:
        sig_result = verify_signature(manifest, public_key)
    except VerificationError as e:
        if args.json:
            print(json.dumps({"valid": False, "error": "signature", "message": str(e)}, indent=2))
        elif not args.quiet:
            print(f"\n[FAILED] Signature verification failed", file=sys.stderr)
            print(f"  {e}", file=sys.stderr)
        sys.exit(1)

    # Step 3: Check rollback protection
    try:
        check_rollback_protection(manifest, args.current_version)
    except RollbackError as e:
        if args.json:
            print(json.dumps({"valid": False, "error": "rollback", "message": str(e)}, indent=2))
        elif not args.quiet:
            print(f"\n[BLOCKED] Rollback protection triggered", file=sys.stderr)
            print(f"  {e}", file=sys.stderr)
        sys.exit(3)

    # Step 4: Check device compatibility
    try:
        check_device_compatibility(manifest, args.device_type, args.device_model)
    except CompatibilityError as e:
        if args.json:
            print(json.dumps({"valid": False, "error": "compatibility", "message": str(e)}, indent=2))
        elif not args.quiet:
            print(f"\n[BLOCKED] Device compatibility error", file=sys.stderr)
            print(f"  {e}", file=sys.stderr)
        sys.exit(4)

    # Success
    metadata = manifest.get("metadata", {})
    result = {
        "valid": True,
        "firmware": str(args.firmware),
        "version": metadata.get("version"),
        "version_code": metadata.get("version_code"),
        "device_type": metadata.get("device_type"),
        "algorithm": sig_result["algorithm"],
        "security_level": sig_result["security_level"],
        "timestamp": sig_result["timestamp"],
        "signer": manifest.get("signer", {}),
    }

    if args.json:
        print(json.dumps(result, indent=2))
    elif not args.quiet:
        print(f"\n[OK] Firmware verification successful!")
        print(f"")
        print(f"  Firmware:       {args.firmware}")
        print(f"  Version:        {metadata.get('version')} (code: {metadata.get('version_code')})")
        print(f"  Device Type:    {metadata.get('device_type')}")
        print(f"  Algorithm:      {sig_result['algorithm']}")
        print(f"  Security Level: {sig_result['security_level']}")
        print(f"  Signed:         {format_timestamp(sig_result['timestamp'])}")
        if manifest.get("signer", {}).get("name"):
            print(f"  Signer:         {manifest['signer']['name']}")
        print(f"")
        print(f"  Status: SAFE TO INSTALL")

    sys.exit(0)


if __name__ == "__main__":
    main()

File: dsa/examples/use-cases/firmware-signing/verify_firmware.cpp
/**
 * Post-Quantum Firmware Verification Tool (C++)
 *
 * Verifies firmware signatures, including rollback protection and
 * device compatibility checks for secure boot and OTA updates.
 *
 * Usage:
 *   ./verify_firmware <public_key> <firmware> [options]
 *
 * Options:
 *   --manifest <file>        Manifest file (default: <firmware>.fwsig)
 *   --current-version <code> Current version code for rollback check
 *   --device-type <type>     Device type for compatibility check
 *   --device-model <model>   Device model for compatibility check
 *   --quiet                  Exit code only
 *   --json                   Output result as JSON
 *
 * Exit codes:
 *   0 - Verification successful, safe to install
 *   1 - Signature or integrity verification failed
 *   2 - Input error
 *   3 - Rollback protection triggered
 *   4 - Device compatibility error
 *
 * Example:
 *   ./verify_firmware keys/fw_public.key firmware.bin --current-version 2000000
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <regex>

#include "mldsa/mldsa.hpp"
#include "slhdsa/slh_dsa.hpp"

#include <openssl/sha.h>

// Exit codes
constexpr int EXIT_VALID = 0;
constexpr int EXIT_INVALID = 1;
constexpr int EXIT_ERROR = 2;
constexpr int EXIT_ROLLBACK = 3;
constexpr int EXIT_COMPATIBILITY = 4;

// Read binary file
std::vector<uint8_t> read_file(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::vector<uint8_t>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Read text file
std::string read_text_file(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::string(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Convert hex string to bytes
std::vector<uint8_t> from_hex(const std::string& hex) {
    std::vector<uint8_t> bytes;
    for (size_t i = 0; i < hex.length(); i += 2) {
        bytes.push_back(static_cast<uint8_t>(std::stoi(hex.substr(i, 2), nullptr, 16)));
    }
    return bytes;
}

// Convert bytes to hex string
std::string to_hex(const std::vector<uint8_t>& data) {
    std::ostringstream oss;
    for (uint8_t b : data) {
        oss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(b);
    }
    return oss.str();
}

// Compute SHA-256 hash
std::vector<uint8_t> sha256(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA256_DIGEST_LENGTH);
    SHA256(data.data(), data.size(), hash.data());
    return hash;
}

// Compute SHA-512 hash
std::vector<uint8_t> sha512(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA512_DIGEST_LENGTH);
    SHA512(data.data(), data.size(), hash.data());
    return hash;
}

// Extract filename from path
std::string get_filename(const std::string& path) {
    size_t pos = path.find_last_of("/\\");
    return (pos == std::string::npos) ? path : path.substr(pos + 1);
}

// Simple JSON value extraction
std::string json_get_string(const std::string& json, const std::string& key) {
    std::regex pattern("\"" + key + "\"\\s*:\\s*\"([^\"]+)\"");
    std::smatch match;
    if (std::regex_search(json, match, pattern)) {
        return match[1].str();
    }
    return "";
}

int json_get_int(const std::string& json, const std::string& key) {
    std::regex pattern("\"" + key + "\"\\s*:\\s*(-?\\d+)");
    std::smatch match;
    if (std::regex_search(json, match, pattern)) {
        return std::stoi(match[1].str());
    }
    return 0;
}

// Extract array from JSON
std::vector<std::string> json_get_array(const std::string& json, const std::string& key) {
    std::vector<std::string> result;
    std::regex pattern("\"" + key + "\"\\s*:\\s*\\[([^\\]]+)\\]");
    std::smatch match;
    if (std::regex_search(json, match, pattern)) {
        std::string array_content = match[1].str();
        std::regex item_pattern("\"([^\"]+)\"");
        auto begin = std::sregex_iterator(array_content.begin(), array_content.end(), item_pattern);
        auto end = std::sregex_iterator();
        for (auto it = begin; it != end; ++it) {
            result.push_back((*it)[1].str());
        }
    }
    return result;
}

// JSON escape string
std::string json_escape(const std::string& s) {
    std::string result;
    for (char c : s) {
        switch (c) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            default: result += c;
        }
    }
    return result;
}

// Detect algorithm from public key size
std::string detect_algorithm(size_t pk_size) {
    switch (pk_size) {
        case 1312: return "mldsa44";
        case 1952: return "mldsa65";
        case 2592: return "mldsa87";
        case 32: return "slh-shake-128f";
        case 48: return "slh-shake-192f";
        case 64: return "slh-shake-256f";
        default:
            throw std::invalid_argument("Cannot detect algorithm from key size: " +
                                        std::to_string(pk_size));
    }
}

// Verify function template for ML-DSA
template<typename DSA>
bool verify_mldsa(const std::vector<uint8_t>& pk,
                  const std::vector<uint8_t>& message,
                  const std::vector<uint8_t>& signature,
                  const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.verify(pk, message, signature, ctx);
}

// Verify function template for SLH-DSA
template<typename DSA>
bool verify_slhdsa(const std::vector<uint8_t>& pk,
                   const std::vector<uint8_t>& message,
                   const std::vector<uint8_t>& signature,
                   const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.verify(pk, message, signature, ctx);
}

struct VerificationResult {
    bool valid = false;
    bool hash_valid = false;
    bool size_valid = false;
    bool signature_valid = false;
    bool rollback_ok = true;
    bool compatibility_ok = true;
    std::string algorithm;
    std::string filename;
    size_t file_size = 0;
    std::string version;
    int version_code = 0;
    std::string device_type;
    std::string timestamp;
    std::string error;
    int exit_code = 0;
};

void print_usage() {
    std::cout << "Post-Quantum Firmware Verification Tool (C++)\n";
    std::cout << std::string(60, '=') << "\n\n";
    std::cout << "Usage: verify_firmware <public_key> <firmware> [options]\n\n";
    std::cout << "Required:\n";
    std::cout << "  <public_key>            Public key file\n";
    std::cout << "  <firmware>              Firmware binary to verify\n\n";
    std::cout << "Options:\n";
    std::cout << "  --manifest <file>       Manifest file (default: <firmware>.fwsig)\n";
    std::cout << "  --current-version <code> Current version code for rollback check\n";
    std::cout << "  --device-type <type>    Device type for compatibility check\n";
    std::cout << "  --device-model <model>  Device model for compatibility check\n";
    std::cout << "  --quiet                 Exit code only\n";
    std::cout << "  --json                  Output result as JSON\n";
    std::cout << "  --help                  Show this help\n\n";
    std::cout << "Exit Codes:\n";
    std::cout << "  0 - Verification successful, safe to install\n";
    std::cout << "  1 - Signature or integrity verification failed\n";
    std::cout << "  2 - Input error\n";
    std::cout << "  3 - Rollback protection triggered\n";
    std::cout << "  4 - Device compatibility error\n\n";
    std::cout << "Example:\n";
    std::cout << "  ./verify_firmware keys/fw_public.key firmware.bin --current-version 2000000\n";
}

void print_json_result(const VerificationResult& result) {
    std::cout << "{\n";
    std::cout << "  \"valid\": " << (result.valid ? "true" : "false") << ",\n";
    std::cout << "  \"checks\": {\n";
    std::cout << "    \"hash\": " << (result.hash_valid ? "true" : "false") << ",\n";
    std::cout << "    \"size\": " << (result.size_valid ? "true" : "false") << ",\n";
    std::cout << "    \"signature\": " << (result.signature_valid ? "true" : "false") << ",\n";
    std::cout << "    \"rollback\": " << (result.rollback_ok ? "true" : "false") << ",\n";
    std::cout << "    \"compatibility\": " << (result.compatibility_ok ? "true" : "false") << "\n";
    std::cout << "  },\n";
    std::cout << "  \"firmware\": {\n";
    std::cout << "    \"name\": \"" << json_escape(result.filename) << "\",\n";
    std::cout << "    \"size\": " << result.file_size << ",\n";
    std::cout << "    \"version\": \"" << json_escape(result.version) << "\",\n";
    std::cout << "    \"version_code\": " << result.version_code << "\n";
    std::cout << "  },\n";
    std::cout << "  \"device_type\": \"" << json_escape(result.device_type) << "\",\n";
    std::cout << "  \"algorithm\": \"" << result.algorithm << "\",\n";
    std::cout << "  \"timestamp\": \"" << result.timestamp << "\"";
    if (!result.error.empty()) {
        std::cout << ",\n  \"error\": \"" << json_escape(result.error) << "\"";
    }
    std::cout << "\n}\n";
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        print_usage();
        return (argc == 1) ? 0 : EXIT_ERROR;
    }

    std::string pk_file;
    std::string fw_file;
    std::string manifest_file;
    int current_version = -1;
    std::string device_type;
    std::string device_model;
    bool quiet = false;
    bool json_output = false;

    // Parse arguments
    int i = 1;
    while (i < argc) {
        std::string arg = argv[i];

        if (arg == "--manifest" && i + 1 < argc) {
            manifest_file = argv[++i];
        } else if (arg == "--current-version" && i + 1 < argc) {
            current_version = std::stoi(argv[++i]);
        } else if (arg == "--device-type" && i + 1 < argc) {
            device_type = argv[++i];
        } else if (arg == "--device-model" && i + 1 < argc) {
            device_model = argv[++i];
        } else if (arg == "--quiet" || arg == "-q") {
            quiet = true;
        } else if (arg == "--json") {
            json_output = true;
        } else if (arg == "--help" || arg == "-h") {
            print_usage();
            return 0;
        } else if (arg[0] != '-') {
            if (pk_file.empty()) {
                pk_file = arg;
            } else if (fw_file.empty()) {
                fw_file = arg;
            }
        } else {
            std::cerr << "Unknown option: " << arg << "\n";
            return EXIT_ERROR;
        }
        ++i;
    }

    if (pk_file.empty() || fw_file.empty()) {
        std::cerr << "Error: Public key and firmware file are required\n";
        return EXIT_ERROR;
    }

    if (manifest_file.empty()) {
        manifest_file = fw_file + ".fwsig";
    }

    VerificationResult result;

    try {
        // Read public key
        if (!quiet && !json_output) std::cerr << "Reading public key: " << pk_file << "\n";
        auto pk = read_file(pk_file);

        // Read firmware
        if (!quiet && !json_output) std::cerr << "Reading firmware: " << fw_file << "\n";
        auto fw_data = read_file(fw_file);
        result.file_size = fw_data.size();
        result.filename = get_filename(fw_file);

        // Read manifest
        if (!quiet && !json_output) std::cerr << "Reading manifest: " << manifest_file << "\n";
        std::string manifest = read_text_file(manifest_file);

        // Extract values from manifest
        std::string alg_id = json_get_string(manifest, "id");
        std::string sig_hex = json_get_string(manifest, "value");
        std::string sha256_hex = json_get_string(manifest, "sha256");
        std::string sha512_hex = json_get_string(manifest, "sha512");
        std::string version = json_get_string(manifest, "version");
        int version_code = json_get_int(manifest, "version_code");
        std::string manifest_device_type = json_get_string(manifest, "device_type");
        std::string timestamp = json_get_string(manifest, "timestamp");
        size_t expected_size = json_get_int(manifest, "size");
        std::vector<std::string> compatibility = json_get_array(manifest, "compatibility");

        result.algorithm = alg_id;
        result.version = version;
        result.version_code = version_code;
        result.device_type = manifest_device_type;
        result.timestamp = timestamp;

        // Detect algorithm from key if not in manifest
        if (alg_id.empty()) {
            alg_id = detect_algorithm(pk.size());
            result.algorithm = alg_id;
        }

        // Verify file size
        result.size_valid = (fw_data.size() == expected_size);

        // Verify hashes
        auto computed_sha256 = sha256(fw_data);
        auto computed_sha512 = sha512(fw_data);
        std::string computed_sha256_hex = to_hex(computed_sha256);
        std::string computed_sha512_hex = to_hex(computed_sha512);

        result.hash_valid = (computed_sha256_hex == sha256_hex);
        if (!sha512_hex.empty()) {
            result.hash_valid = result.hash_valid && (computed_sha512_hex == sha512_hex);
        }

        // Check rollback protection
        if (current_version >= 0) {
            if (version_code < current_version) {
                result.rollback_ok = false;
                result.error = "Rollback blocked: firmware version " + std::to_string(version_code) +
                              " is older than current " + std::to_string(current_version);
            }
        }

        // Check device compatibility
        if (!device_type.empty()) {
            if (device_type != manifest_device_type) {
                result.compatibility_ok = false;
                result.error = "Device type mismatch: expected " + manifest_device_type +
                              ", got " + device_type;
            }
        }

        if (!device_model.empty() && !compatibility.empty()) {
            bool model_found = std::find(compatibility.begin(), compatibility.end(), device_model) != compatibility.end();
            if (!model_found) {
                result.compatibility_ok = false;
                result.error = "Device model " + device_model + " not in compatibility list";
            }
        }

        // Build message to verify
        std::string fw_name = json_get_string(manifest, "name");
        if (fw_name.empty()) fw_name = result.filename;

        std::ostringstream msg_stream;
        msg_stream << "{\"firmware_hash\":\"" << sha256_hex << "\",";
        msg_stream << "\"firmware_name\":\"" << fw_name << "\",";
        msg_stream << "\"firmware_size\":" << expected_size << ",";
        msg_stream << "\"version\":\"" << version << "\",";
        msg_stream << "\"version_code\":" << version_code << ",";
        msg_stream << "\"device_type\":\"" << manifest_device_type << "\",";
        msg_stream << "\"timestamp\":\"" << timestamp << "\"}";
        std::string msg_str = msg_stream.str();
        std::vector<uint8_t> message(msg_str.begin(), msg_str.end());

        // Decode signature
        std::vector<uint8_t> signature = from_hex(sig_hex);

        // Firmware context
        std::vector<uint8_t> ctx = {'f', 'i', 'r', 'm', 'w', 'a', 'r', 'e'};

        // Verify signature
        if (!quiet && !json_output) std::cerr << "Verifying signature with " << alg_id << "...\n";

        if (alg_id == "mldsa44") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA44>(pk, message, signature, ctx);
        } else if (alg_id == "mldsa65") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA65>(pk, message, signature, ctx);
        } else if (alg_id == "mldsa87") {
            result.signature_valid = verify_mldsa<mldsa::MLDSA87>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-128f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_128f>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-192f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_192f>(pk, message, signature, ctx);
        } else if (alg_id == "slh-shake-256f") {
            result.signature_valid = verify_slhdsa<slhdsa::SLHDSA_SHAKE_256f>(pk, message, signature, ctx);
        } else {
            throw std::invalid_argument("Unsupported algorithm: " + alg_id);
        }

        // Determine final result and exit code
        if (!result.rollback_ok) {
            result.exit_code = EXIT_ROLLBACK;
        } else if (!result.compatibility_ok) {
            result.exit_code = EXIT_COMPATIBILITY;
        } else if (!result.hash_valid || !result.size_valid || !result.signature_valid) {
            result.exit_code = EXIT_INVALID;
        } else {
            result.exit_code = EXIT_VALID;
        }

        result.valid = (result.exit_code == EXIT_VALID);

    } catch (const std::exception& e) {
        result.error = e.what();
        result.exit_code = EXIT_ERROR;
        if (json_output) {
            print_json_result(result);
        } else if (!quiet) {
            std::cerr << "Error: " << e.what() << "\n";
        }
        return EXIT_ERROR;
    }

    // Output result
    if (json_output) {
        print_json_result(result);
    } else if (!quiet) {
        std::cout << "\n" << std::string(50, '=') << "\n";
        std::cout << "  Firmware Verification Result\n";
        std::cout << std::string(50, '=') << "\n\n";

        std::cout << "Firmware:     " << result.filename << "\n";
        std::cout << "Size:         " << result.file_size << " bytes\n";
        std::cout << "Version:      " << result.version << " (code: " << result.version_code << ")\n";
        std::cout << "Device Type:  " << result.device_type << "\n";
        std::cout << "Algorithm:    " << result.algorithm << "\n";
        std::cout << "Timestamp:    " << result.timestamp << "\n";

        std::cout << "\nChecks:\n";
        std::cout << "  Hash:          " << (result.hash_valid ? "PASS" : "FAIL") << "\n";
        std::cout << "  Size:          " << (result.size_valid ? "PASS" : "FAIL") << "\n";
        std::cout << "  Signature:     " << (result.signature_valid ? "PASS" : "FAIL") << "\n";
        std::cout << "  Rollback:      " << (result.rollback_ok ? "PASS" : "BLOCKED") << "\n";
        std::cout << "  Compatibility: " << (result.compatibility_ok ? "PASS" : "FAIL") << "\n";

        if (!result.error.empty()) {
            std::cout << "\nError: " << result.error << "\n";
        }

        std::cout << "\n" << std::string(50, '=') << "\n";
        if (result.valid) {
            std::cout << "  VERIFICATION SUCCESSFUL - SAFE TO INSTALL\n";
        } else {
            std::cout << "  VERIFICATION FAILED - DO NOT INSTALL\n";
        }
        std::cout << std::string(50, '=') << "\n";
    }

    return result.exit_code;
}
